### ISSUE TITLE
[possible bug?] GridGenerator::hyper_ball does not produce circular edges at the boundary

### ISSUE THREAD

**Issue by krishnakumarg1984** [dealii/dealii#9296]
Apologies if this issue report seems a bit far-fetched. As the issue subject says, `GridGenerator::hyper_ball`  does not seem to be producing circular boundaries

I was trying out step-6 tutorial, and noticed this in the output mesh. At first, I thought I made an error (see also my recently merged documentation-related PR #9295), and tried attaching a `SphericalManifold` to the boundary. Despite this, the mesh produced had straight edges. Before posting the image here and filing an issue against step#6, I thought of going back to step-1 and using the `hyper_ball` mesh without any global refinement.

```
#include <deal.II/grid/tria.h>
#include <deal.II/grid/tria_accessor.h>
#include <deal.II/grid/tria_iterator.h>
#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/grid_out.h>

#include <iostream>
#include <fstream>
#include <cmath>

using namespace dealii;

void first_grid()
{
  Triangulation<3> triangulation;
  GridGenerator::hyper_ball(triangulation);
  std::ofstream out("grid-1.svg");
  GridOut       grid_out;
  grid_out.write_svg(triangulation, out);
  std::cout << "Grid written to grid-1.svg" << std::endl;
}

int main()
{
  first_grid();
}
```

This produces the following mesh.
<img width="665" alt="Screenshot 2020-01-12 at 21 16 34" src="https://user-images.githubusercontent.com/15239509/72225745-dbec3d80-3580-11ea-815d-a71a2accef4a.png">

According to the documentation, 
>Initialize the given triangulation with several coarse mesh cells that cover a hyperball, i.e. a circle or a ball around center with given radius.
>.....
By default, the manifold_id is set to 0 on the boundary faces, 1 on the boundary cells, and types::flat_manifold_id on the central cell and on internal faces.
>
>A SphericalManifold is attached by default to the boundary faces for correct placement of boundary vertices upon refinement and to be able to use higher order mappings. 

Clearly, the desired behaviour is not currently happening. Is it something I am doing incorrectly, or is this perhaps a bug in the library?

**Comment by blaisb** [comment]
In the current case, you have an insufficient number of elements and you are displaying first order elements (linear).
There are two ways you can fix your issue.
1- Increase the number of elements. As you refine the triangulation, it will tend towards a real circle that is discretized using linear elements.
2- You can use high-order element display within paraview and use high order mapping. This page documents this feature in a very clear fashion : https://github.com/dealii/dealii/wiki/Notes-on-visualizing-high-order-output

Remember, by default, paraview was made to visualize linear elements. Consequently, all edges will appear to be lines as in regular second-order accurate methods.

**Comment by krishnakumarg1984** [comment]
@blaisb Thank you. That helps, and explains most of my issues.

However, there still remains an unexplained mystery.  How were the nice circular meshes written as .eps files by step-6 produced  (see (here)[https://dealii.org/developer/doxygen/deal.II/step_6.html#Results])? At least in the first and second refinement cycle, we get jagged edges, and there is no way, that the figures shown here could be reproduced with the code in step-6.

**Comment by blaisb** [comment]
This remains a guess, but I presume that when you use write_eps with higher order (say >Q1) and you have manifolds associated with the boundaries (as is the case by default for the hyper_ball, then the write_eps takes it into account. In the image you showed above, you have only Q1 elements in your dof_handler right? I suggest you try it again with higher order.

**Comment by drwells** [comment]
Thanks for finding this problem with step-6. A few quick comments:

1. Regarding the SVG output: our SVG writer does not know how to print curved cells, only quadrilaterals. Hence the output is always incorrect in the sense that it ignores the available manifold information. Other output writers (like EPS) have ways to fix this.

2. The element space is not related to this problem: in fact, we only write out the Triangulation and never read anything from a `FiniteElement` or `DoFHandler`. The only relevant objects here are `Triangulation`, `Manifold`, and (to get curved output) `Mapping`. 

3. This is definitely a problem with step-6. By default `write_eps` ignores manifold information on boundary faces (it uses `MappingQGeneric(1)`). We need both to provide a higher-order mapping to `write_eps` *and* also tell it to plot more points on the boundary (since EPS just plots line segments we make it look curved by using a huge number of line segments).
I can reproduce the pictures by changing the code in step-6 to the following:
```cpp
  {
    GridOut       grid_out;
    GridOutFlags::Eps<dim> eps_flags;
    eps_flags.n_boundary_face_points = 10;
    grid_out.set_flags(eps_flags);
    std::ofstream output("grid-" + std::to_string(cycle) + ".eps");
    MappingQ<dim> mapping(3);
    grid_out.write_eps(triangulation, output, &mapping);
  }
```
@krishnakumarg1984 would you be willing to update step-6 with this code and an explanation?

**Comment by blaisb** [comment]
> Thanks for finding this problem with step-6. A few quick comments:
> 
> 1. Regarding the SVG output: our SVG writer does not know how to print curved cells, only quadrilaterals. Hence the output is always incorrect in the sense that it ignores the available manifold information. Other output writers (like EPS) have ways to fix this.
> 2. The element space is not related to this problem: in fact, we only write out the `Triangulation` and never read anything from a `FiniteElement` or `DoFHandler`. The only relevant objects here are `Triangulation`, `Manifold`, and (to get curved output) `Mapping`.
> 3. This is definitely a problem with step-6. By default `write_eps` ignores manifold information on boundary faces (it uses `MappingQGeneric(1)`). We need both to provide a higher-order mapping to `write_eps` _and_ also tell it to plot more points on the boundary (since EPS just plots line segments we make it look curved by using a huge number of line segments).
>    I can reproduce the pictures by changing the code in step-6 to the following:
> 
> ```c++
>   {
>     GridOut       grid_out;
>     GridOutFlags::Eps<dim> eps_flags;
>     eps_flags.n_boundary_face_points = 10;
>     grid_out.set_flags(eps_flags);
>     std::ofstream output("grid-" + std::to_string(cycle) + ".eps");
>     MappingQ<dim> mapping(3);
>     grid_out.write_eps(triangulation, output, &mapping);
>   }
> ```
> 
> @krishnakumarg1984 would you be willing to update step-6 with this code and an explanation?

Thanks, that is enlightening. I was surprised that something could  yield curved boundaries without having been provided a non-Q1 mapping.

**Comment by krishnakumarg1984** [comment]
@blaisb 

>In the image you showed above, you have only Q1 elements in your dof_handler right? 

I don't even have a dof_handler object in the code. It is purely the visualisation of the triangulation as done in step-1. 

>but I presume that when you use write_eps with higher order (say >Q1) and you have manifolds associated with the boundaries (as is the case by default for the hyper_ball, then the write_eps takes it into account.

Step-6 by default uses bi-quadratic elements. If you run the Step-6 tutorial on your machine, you will __not__ get the circular shape of the mesh for the 1st few refinement cycles.

**Comment by krishnakumarg1984** [comment]
>Regarding the SVG output: our SVG writer does not know how to print curved cells, only quadrilaterals. Hence, the output is always incorrect in the sense that it ignores the available manifold information. Other output writers (like EPS) have ways to fix this.

This sounds like a problem. Newcomers to the library expect to visualize the mesh the way it is.  See #5982 and #8447 on how this affects the tutorials.

>By default write_eps ignores manifold information on boundary faces (it uses MappingQGeneric(1)). 

This information is not revealed to the user in the documentation of the [`write_eps`](https://dealii.org/developer/doxygen/deal.II/classGridOut.html#a7e31485bb55254bfa5cd3ab55f8d1748) function. 

>We need both to provide a higher-order mapping to write_eps and also tell it to plot more points on the boundary (since EPS just plots line segments we make it look curved by using a huge number of line segments).

This seems like a bit clunky way of just getting to look at the grid. Is there any other writer that can visualize the grid correctly without so many control knobs?  For instance, how do we know `10` is the right number to use, and why is the `3rd` order mapping the optimal one?

>would you be willing to update step-6 with this code and an explanation?

Yes, I am happy to do this after a bit more clarity emerges from the discussion. I am still puzzled by the couple of points above. A simple writer that can automagically visualize curved domains is what the tutorials must point to.

**Comment by drwells** [comment]
> In face, the GridOutFlags::Eps() class' documentation is unbelievably misleading.

This looks like an issue with the way doxygen picks things up: in particular, 'below' refers to what is done in the header. Would you be willing to put together a fix?

**Comment by drwells** [comment]
That much, at least, is independent of the output of curved meshes in the tutorial programs.

**Comment by krishnakumarg1984** [comment]
>This looks like an issue with the way doxygen picks things up: in particular, 'below' refers to what is done in the header. Would you be willing to put together a fix?

I don't exactly understand the fix needed here.

Also, please address the other aspects I have raised in this issue.

**Comment by drwells** [comment]
> This sounds like a problem. Newcomers to the library expect to visualize the mesh the way it is.

I agree with what @bangerth wrote in the other issue about this: the *default* is to not use higher-order mappings so any choice is misleading as to what is actually computed since the computational choices and geometry description are independent. Our documentation on this point could be improved but the way things are is a fundamental choice in the library. Additionally, not all output formats support more than a linear representation so anything we do is going to be a bit of a hack.

> This information is not revealed to the user in the documentation of the write_eps function.

The function documentation states "mapping is a pointer to a mapping used for the transformation of cells at the boundary. If zero, then use standard Q1 mapping."

> This seems like a bit clunky way of just getting to see the grid. Is there any other writer that can visualize the grid correctly without so many control knobs? For instance, how do we know 10 is the right number to use, and why is the 3rd order mapping the optimal one?

Looking for a right answer here is like looking for a correct answer to 'what order finite element space should I use' - its always going to depend on more things than we can anticipate. Like elsewhere in the library, we might declare the geometry to have certain properties, but when we compute we have to explicitly choose a polynomial approximation. This is part of the reason why we compare multiple isoparametric mappings in step-10.

> In face, the GridOutFlags::Eps() class' documentation is unbelievably misleading.

The documentation for this class isn't rendered perfectly due to the way we use inheritance and templates together. Try looking at https://dealii.org/developer/doxygen/deal.II/structGridOutFlags_1_1EpsFlagsBase.html instead.

> Even after including the code you provided, only the boundary cells are visualized as curved. However, with the code in the extensions section for attaching a spherical manifold to all cells and setting its id to 0, the straight edges of the central cell remain. Therefore, we cannot reproduce the fully circular looking mesh graphic given in the extensions section.

IIRC (see #6381, its been awhile though) I generated those plots with the gnuplot writer and then saved them to SVG. That's why the interior cells appear to be curved. I honestly don't remember why we kept that as EPS instead of gnuplot (with further plotting directions).
